### ISSUE TITLE
[security] ss-server opens access to local and/or firewalled services to any remote shadowsocks user

### ISSUE THREAD

**Issue by sleeksorrow** [shadowsocks/shadowsocks-libev#2265]
### What version of shadowsocks-libev are you using?
3.1.3

### What operating system are you using?
Ubuntu 18.04

### What did you do?
I'm having a shadowsocks server running on a host where also other services are running, that are firewalled or only listening on localhost. In this example I have one MariaDB running on my server listening only on 127.0.0.1:3306
Now I'm using a clean client system (Ubuntu 18.10) with ss-local to access the MariaDB on the server through the shadowsocks connection.
`nc -x 127.0.01:1090 127.0.0.1 3306`

### What did you expect to see?
Being blocked by shadowsocks server

### What did you see instead?
Being able to access the MariaDB on the remote server
```
nc -x 127.0.01:1090 127.0.0.1 3306
5.5.5-10.1.34-MariaDB-0ubuntu0.18.04.1
```

### What is your config in detail (with all sensitive info masked)?
Client:
```
{
    "server":["XXXXXXXXXXXXX"],
    "server_port":443,
    "local_port":1090,
    "password":"XXXXXXX",
    "timeout":360,
    "method":"chacha20-ietf-poly1305",
    "prefer_ipv6": true,
    "ipv6_first": true,
    "fast_open": true,
    "reuse_port": true,
    "no_delay": true
}
```
Server:
```
{
    "server":["XXXXXXXXXXXX"],
    "server_port":443,
    "local_port":1090,
    "password":"XXXXXXXXXX",
    "timeout":360,
    "method":"chacha20-ietf-poly1305",
    "prefer_ipv6": true,
    "ipv6_first": true,
    "mode": "tcp_and_udp",
    "no_delay": true,
    "fast_open": true,
    "reuse_port": true
}

```

This behavior means that with the shadowsocks server running on the host, all services which are only listening on localhost or which are firewalled are now accessible to anybody knowing the shadowsocks port and password. Since shadowsocks is meant to be a service that is very easily shared between a loosely defined group of unidentified users, I think this unknowingly opens a lot of attack vectors and that way causes a security threat.

Proposal:
Add an option that takes a list of forbidden destination IP addresses that will be blocked by ss-server, and if unset, forbid all local IPs by default.

**Comment by sleeksorrow** [comment]
Agreed, the acl system looks way more flexible than my proposal. Thanks a lot and sorry I missed that.

The issue is still valid, tho, but I want to change my proposal into having a default acl which is blocking access to all local IPs by default, but granting an override.

**Comment by madeye** [comment]
Try `ssh -D` and `ssh -L` first. If you still think it's a security issue, please report it to OpenSSH project, since shadowsocks follows the behavior of OpenSSH.

In addition, we provided server-side ACL feature.
